testool/src/compiler.rs

```diff
@@ -71,6 +72,139 @@ pub struct Compiler {
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CompilerSettings {
+    optimizer: Optimizer,
+    output_selection: HashMap<String, HashMap<String, Vec<String>>>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Optimizer {
+    enabled: bool,
+    details: HashMap<String, bool>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Source {
+    content: String,
+}
+
+impl CompilerInput {
+    pub fn new_default(language: Language, src: &str) -> Self {
+        let mut sources = HashMap::new();
+        sources.insert(
+            "stdin".to_string(),
+            Source {
+                content: src.to_string(),
+            },
+        );
+        CompilerInput {
+            language,
+            settings: Default::default(),
+            sources,
+        }
+    }
+}
+
+impl Default for CompilerSettings {
+    fn default() -> Self {
+        let mut output_selection = HashMap::new();
+        let mut selection = HashMap::new();
+        selection.insert("*".to_string(), vec!["evm.bytecode".to_string()]);
+        output_selection.insert("*".to_string(), selection);
+        CompilerSettings {
+            optimizer: Default::default(),
+            output_selection,
+        }
+    }
+}
+
+impl Default for Optimizer {
+    fn default() -> Self {
+        let mut details = HashMap::new();
+        details.insert("peephole".to_string(), false);
+        details.insert("inliner".to_string(), false);
+        details.insert("jumpdestRemover".to_string(), false);
+        Optimizer {
+            enabled: false,
+            details,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CompilationResult {
+    contracts: HashMap<String, HashMap<String, Contract>>,
+    errors: Vec<CompilationError>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Contract {
+    evm: Evm,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Evm {
+    bytecode: BytecodeResult,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct BytecodeResult {
+    object: String,
+    opcodes: String,
+    source_map: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CompilationError {
+    component: String,
+    error_code: String,
+    formatted_message: String,
+    message: String,
+    severity: Severity,
+    #[serde(rename = "type")]
+    error_type: String,
+    source_location: Option<SourceLocation>,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "lowercase")]
+enum Severity {
+    Error,
+    Warning,
+    Info,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct SourceLocation {
+    end: i32,
+    file: String,
+    start: i32,
+}
+
 impl Compiler {
```

Comment:
Why is the compiler impl so far from the compiler struct, if this compiler depends now in the new structs, perhaps we should move struct Compiler down, right before impl Compiler?

Reply:
solved